### PR TITLE
Doc system refactor/cleanup

### PR DESCRIPTION
**Pull request by MikeInnes**
Mostly just moving things around, takes advantages of the now-working bootstrap phase, also fixes docs for `@doc`. Once this is in will have a go at modules.

### CODE REVIEW

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -2,6 +2,62 @@
 
 module Docs
 
+"""
+The Docs module provides the `@doc` macro which can be used to set and retreive
```
Comment (joehuchette) [inline base/docs/Docs.jl]: *retrieve

**Comment by tkelman** [comment]
Should still really let CI finish before merging things. And you're missing license headers on the newly added files.

**Comment by one-more-minute** [comment]
Running CI was why I put this up as a PR in the first place. Presumably correcting spelling in a docstring isn't going to break the build, unless we have much bigger issues.

**Comment by tkelman** [comment]
Ah, nevermind, missed that the pre-rebase build did pass.

Still nice to at least give some time for people to check and review things like the license headers, people are distributed across all time zones after all.

**Comment by one-more-minute** [comment]
Well, I guess we'll just have to hope that no one sues us between now and the time I add the headers.

**Comment by tkelman** [comment]
Not the point. The point is code review.

**Comment by StefanKarpinski** [comment]
Given that we really need to get this documentation stuff in better shape asap, I think we can do a little post-merge code review here if it lets @one-more-minute get things he needs for docs merged faster.

**Comment by tkelman** [comment]
Fine, but let's not make post-merge code review end up needing to be reversion again by being too hasty.

**Comment by StefanKarpinski** [comment]
Agreed, we'll try not to do that.

**Comment by StefanKarpinski** [comment]
Ok, ok. Be cool, folks. The headers aren't a big deal (I didn't even want them). They're not to prevent us from being sued – they're to make it less likely for code to be lifted accidentally from Julia and transplanted into some other code base without complying with the Julia license.

**Comment by tkelman** [comment]
I'm perfectly cool, just advocating for doing things carefully and the right way, same way I do with everyone. License headers aren't a big deal, this was a harmless code reorganization which is almost always fine, and I have nothing at all against @one-more-minute's work (if he's getting that impression then I apologize). But if anyone is going to open a PR for something that's more than just a few lines, merging it an hour after opening doesn't give many people a chance to look at it. We're trying to get a release out, but if we rush everything it will be objectively worse than if we're careful about what needs to be done and weighing the risks vs benefits.

**Pull request merged**